File: 0131-palindrome-partitioning/0131-palindrome-partitioning.cpp
class Solution {
public:
    vector<vector<string>> partition(string s) {
        vector<vector<string>> ans;
        vector<string> path;
        
        calc(s,path,0,ans);
        
        return ans;
    }
    
    void calc(string& s,vector<string>& path,int index,vector<vector<string>>& ans){
        if(index == s.size()){
            ans.push_back(path);
            return;
        }
        
        for(int i=index;i<s.size();i++){
            if(palindrome(s,index,i)){
                path.push_back(s.substr(index,i-index+1));
                calc(s,path,i+1,ans);
                path.pop_back();
            }
        }
    }
    
    bool palindrome(string& s,int start,int end){
        while(start<=end){
            if(s[start++] != s[end--]) return false;
        }
        
        return true;
    }

};

File: 0131-palindrome-partitioning/0131-palindrome-partitioning.py
class Solution:
    def partition(self, s: str) -> List[List[str]]:
        res = []

        def palindrome(s):
            return s == s[::-1] #reverse
        

        def dfs(start,arr):
            if start == len(s):
                res.append(arr.copy())
                return
            

            for end in range(start,len(s)):
                #slicing
                if palindrome(s[start:end+1]):
                    arr.append(s[start:end+1])
                    dfs(end+1,arr)
                    arr.pop() #backtrack and explore other options
                
            
        dfs(0,[])

        return res

        
